### PR TITLE
Lowercases all ireg links

### PR DESCRIPTION
**Pull request by jslay-excella**
Lowercases all ireg links and also allows for lowercase routable links by filtering on case insensitive when building query set for section.

Fixes https://ghe/CFGOV/platform/issues/3852

## Additions

`lower()` in cfgov/regulations3k/views.py and models/pages.py Python code
- interp-{appendix.lower()}
- interp-{section.lower()}
- section_label.lower()

## Changes

- `Interp -> interp` in `test_resolver.py` and `test_views.py`

## How to test this PR
To check for formatting errors
`tox -e lint`

To run all tests:
`tox -e unittest-current`

To only run regulations3k tests:
1. tox -e unittest-current regulations3k.tests.test_resolver.ReferenceResolutionTestCase
2. tox -e unittest-current regulations3k.tests.test_views.RedirectRegulations3kTestCase

## Notes and todos

-

## Checklist

- [X] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [X] Changes are limited to a single goal (no scope creep)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [ ] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

#### Browser testing

Visually tested in the following supported browsers:
- [ ] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge 18 (the last Edge prior to it switching to Chromium)
- [ ] Internet Explorer 11 and 8 (via emulation in 11's dev tools)
- [ ] Safari on iOS
- [ ] Chrome on Android

#### Accessibility

- [ ] Keyboard friendly (navigable with tab, space, enter, arrow keys, etc.)
- [ ] Screen reader friendly
- [ ] Does not introduce new errors or warnings in [WAVE](https://wave.webaim.org/extension/)

#### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Does not introduce new lint warnings
- [ ] Flexible from small to large screens

### CODE REVIEW

**Review by cwdavies — CHANGES_REQUESTED** [review]
There were lint errors that have now been addressed that were found when running `tox -e lint`
```
./cfgov/regulations3k/views.py:212:80: E501 line too long (86 > 79 characters)
./cfgov/regulations3k/views.py:224:80: E501 line too long (85 > 79 characters)
./cfgov/regulations3k/models/pages.py:567:80: E501 line too long (85 > 79 characters)
```
There are Python unit tests in `consumerfinance.gov/cfgov/regulations3k/tests` that were updated due to changing ireg links to lower case. e.g. `test_resolver.py` and `test_views.py`.
A suggestion is to take a look at `def test_redirect_interp_section_past_lowercase` in `test_views.py`

**Comment by cwdavies** [comment]
When running tox you will see which Python unit tests are failing and on which line(s)
```
  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_resolver.py", line 130, in test_get_url_resolver
    self.assertEqual(result, '/reg-landing/1002/Interp-2/#c-Interp')
AssertionError: '/reg-landing/1002/interp-2/#c-Interp' != '/reg-landing/1002/Interp-2/#c-Interp'
- /reg-landing/1002/interp-2/#c-Interp
?                   ^
+ /reg-landing/1002/Interp-2/#c-Interp
?                   ^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 179, in test_interp_section_current
    '/policy-compliance/rulemaking/regulations/1002/Interp-1/')
AssertionError: '/policy-compliance/rulemaking/regulations/1002/interp-1/' != '/policy-compliance/rulemaking/regulations/1002/Interp-1/'
- /policy-compliance/rulemaking/regulations/1002/interp-1/
?                                                ^
+ /policy-compliance/rulemaking/regulations/1002/Interp-1/
?                                                ^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 188, in test_interp_section_no_subpart_with_default_section
    '/policy-compliance/rulemaking/regulations/1005/'
AssertionError: '/policy-compliance/rulemaking/regulations/1005/2013-03-26/interp-2/' != '/policy-compliance/rulemaking/regulations/1005/2013-03-26/Interp-2/'
- /policy-compliance/rulemaking/regulations/1005/2013-03-26/interp-2/
?                                                           ^
+ /policy-compliance/rulemaking/regulations/1005/2013-03-26/Interp-2/
?                                                           ^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 126, in test_redirect_interp_appendix
    '/policy-compliance/rulemaking/regulations/'
AssertionError: '/policy-compliance/rulemaking/regulations/1002/2016-07-11/interp-c/' != '/policy-compliance/rulemaking/regulations/1002/2016-07-11/Interp-C/'
- /policy-compliance/rulemaking/regulations/1002/2016-07-11/interp-c/
?                                                           ^      ^
+ /policy-compliance/rulemaking/regulations/1002/2016-07-11/Interp-C/
?                                                           ^      ^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 135, in test_redirect_interp_appendix_invalid_date
    '/policy-compliance/rulemaking/regulations/1024/Interp-MS/')
AssertionError: '/policy-compliance/rulemaking/regulations/1024/interp-ms/' != '/policy-compliance/rulemaking/regulations/1024/Interp-MS/'
- /policy-compliance/rulemaking/regulations/1024/interp-ms/
?                                                ^      ^^
+ /policy-compliance/rulemaking/regulations/1024/Interp-MS/
?                                                ^      ^^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 143, in test_redirect_interp_intro
    '/policy-compliance/rulemaking/regulations/1002/'
AssertionError: '/policy-compliance/rulemaking/regulations/1002/2016-07-11/h1-interp/' != '/policy-compliance/rulemaking/regulations/1002/2016-07-11/h1-Interp/'
- /policy-compliance/rulemaking/regulations/1002/2016-07-11/h1-interp/
?                                                              ^
+ /policy-compliance/rulemaking/regulations/1002/2016-07-11/h1-Interp/
?                                                              ^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 152, in test_redirect_interp_intro_bad_version
    '/policy-compliance/rulemaking/regulations/1030/Interp-0/')
AssertionError: '/policy-compliance/rulemaking/regulations/1030/interp-0/' != '/policy-compliance/rulemaking/regulations/1030/Interp-0/'
- /policy-compliance/rulemaking/regulations/1030/interp-0/
?                                                ^
+ /policy-compliance/rulemaking/regulations/1030/Interp-0/
?                                                ^

  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 160, in test_redirect_interp_section_past
    '/policy-compliance/rulemaking/regulations/1002/'
AssertionError: '/policy-compliance/rulemaking/regulations/1002/2016-07-11/interp-1/' != '/policy-compliance/rulemaking/regulations/1002/2016-07-11/Interp-1/'
- /policy-compliance/rulemaking/regulations/1002/2016-07-11/interp-1/
?                                                           ^
+ /policy-compliance/rulemaking/regulations/1002/2016-07-11/Interp-1/
?                                                           ^


  File "/Users/DaviesC/Projects/CFPB/consumerfinance.gov/cfgov/regulations3k/tests/test_views.py", line 170, in test_redirect_interp_section_past_lowercase
    '/policy-compliance/rulemaking/regulations/1002/'
AssertionError: '/policy-compliance/rulemaking/regulations/1002/2011-12-30/interp-1/' != '/policy-compliance/rulemaking/regulations/1002/2011-12-30/Interp-1/'
- /policy-compliance/rulemaking/regulations/1002/2011-12-30/interp-1/
?                                                           ^
+ /policy-compliance/rulemaking/regulations/1002/2011-12-30/Interp-1/
?                                                           ^

Ran 1740 tests in 260.824s

FAILED (failures=9)
```

File: cfgov/regulations3k/models/pages.py
Comment (niqjohnson) [inline cfgov/regulations3k/models/pages.py:470]: I'm pretty sure this should be a permanent redirect, but I'll check with our SEO experts to be sure.

```suggestion
            section = section_query.get(label__iexact=section_label)
            if re.search('[A-Z]+', section_label):
                return redirect(
                    self.url + self.reverse_subpage(
                        "index", kwargs=kwargs
                    ) + section_label.lower() + '/',
                    permanent=True
                )
```

**Review by niqjohnson — CHANGES_REQUESTED** [review]
I tested these changes locally both on the front end by browsing through some regs and on the back end by editing some sections and creating some new versions. Generally things are working great; I found a few little things that we should take a look at.

1. It looks like we're using a 302 temporary redirect to redirect uppercase to lowercase URLs. That should probably be a 301 permanent redirect (which I've added as a review suggestion), but I'm going to check with our SEO experts to make sure.
2. The "Appendices" and "Supplement I…" side navigation sections are no longer showing the active navigation item when you're on appendix or interpretation pages. My suggestion to add `.lower()` to pages.py line 570 fixes that for me, but I'm not sure if there's a better way to do that.
3. The previous and next links to appendix and interpretation pages that appear at the bottom of pages still use the capitalized URLs. I was able to fix that by adding `.lower()` to the `next_url` and `previous_url` definitions in pages.py lines 503 and 505, but I'm sure there's a better way to do that.
4. Links in search results (for example, /rules-policy/regulations/search-regulations/results/?regs=1026&q=foreign) are still using the capitalized URLs.

For all the links that are still pointing to the capitalized URLs, the redirect works great, but we should update the links to point to the proper lowercase URLs so we're not doing a bunch of redirects.

Also, definitely a suggestion and not a bug, but since we're using `section.label.lower()` lots of places now, I wonder if it's worth defining a new context variable for that. Something like `section.url_path` or something? Naming is hard…

File: cfgov/regulations3k/models/django.py
Comment (willbarton) [inline cfgov/regulations3k/models/django.py:244]: I think this is fine as an overall approach, with one caveat.

I waffled a bit on whether this should be something like a template tag or a property on the model. I think a property on the model makes more sense than a template tag if it's a [`cached_property`](https://docs.djangoproject.com/en/3.2/ref/utils/#django.utils.functional.cached_property).

**Review by niqjohnson — CHANGES_REQUESTED** [review]
I found a couple more places where we're still using the capitalized URLs that are over my head for suggesting a fix, so I'll note them here and leave the actual fixing in more capable hands.

First, the previous and next links at the bottom of each page are still linking to the uppercase URLs:

prev-next-links

Second, links in search results are also still using the uppercase URLs:

search-links

I *think* those are the only places where the capitalization remains, but I'll keep poking around to see if I can uncover any more.

File: cfgov/regulations3k/models/pages.py
Comment (Scotchester) [inline cfgov/regulations3k/models/pages.py]: Don't we still need this for the secondary nav items (in the left sidebar)?

Comment (jslay-excella) [reply cfgov/regulations3k/models/pages.py]: no, as `get_section_url` now does the `lower()`

Comment (Scotchester) [reply cfgov/regulations3k/models/pages.py]: Ah, and `get_section_url` is also called in generating the next/previous links?

Comment (jslay-excella) [reply cfgov/regulations3k/models/pages.py]: correct

File: cfgov/regulations3k/models/pages.py
Comment (niqjohnson) [reply cfgov/regulations3k/models/pages.py]: This one is fixed, active nav items are showing as active again

File: cfgov/regulations3k/models/pages.py
Comment (niqjohnson) [reply cfgov/regulations3k/models/pages.py]: Fixed, I'm seeing a 301 redirect for uppercase-to-lowercase URLs in Chrome's dev tools.

**Review by niqjohnson — CHANGES_REQUESTED** [review]
@jslay-excella, this is looking great, all of the uppercase URLs we noticed before are lowercase now. But I keep finding stray uppercase URLs tucked into the dark corners of iRegs. I *think* the last one is the URL the "view live/view draft" button points to when you're editing the regulation in Wagtail (I can show you how to get there if you haven't run across that screen yet):

draft-urls

That URL is set in [cfgov/regulations3k/wagtail_hooks.py#L60](https://github.com/cfpb/consumerfinance.gov/blob/main/cfgov/regulations3k/wagtail_hooks.py#L60); looks like a simple `lower()` in the right place does the trick there, but there might be a better way to do it.

**Review by niqjohnson — APPROVED** [review]
Nice work, @jslay-excella, we've zapped all of the uppercase URLs that I can find. Everything in Wagtail and on the front end is looking good to me now, so I'm going to mark this as approved on my end. Do you want to wait until @higs4281 can give the code a look before merging it, or are we feeling pretty comfortable with where it is?

**Comment by jslay-excella** [comment]
Unless there is a pressing matter to get this deployed, should probably wait for @higs4281 approval as well.

**Review by higs4281 — CHANGES_REQUESTED** [review]
I found one spot that was letting certain appendix uppercase values leak into URLs.

With that, this looks ready to go.

File: cfgov/regulations3k/models/pages.py
Comment (higs4281) [reply cfgov/regulations3k/models/pages.py:128]: Unrelated to this PR, but we should probably rebuild the regulations3k index after this goes in.

**Comment by higs4281** [comment]
> I found one spot that was letting certain appendix uppercase values leak into URLs.


On second thought, we should leave uppercase values after the hash alone. So pls ignore the suggestion.

**Review by higs4281 — APPROVED** [review]
I withdrew my suggestion; I think this is good to go.

**Review by willbarton — APPROVED** [review]
Ran the Cypress tests locally and they all pass. We'll deal with the Jenkins failures in a separate PR.

This looks good to me!

**Pull request merged**
